### ISSUE TITLE
Setting initial state in server-side rendering

### ISSUE THREAD

**Issue by ameensol** [edge/fluxette#9]
You recommend creating a new instance of flux for server-side rendering, but so far I've found that this complicates matters more than it helps. The main issue is making sure that the server route-handler has a reference to the same instance of flux as the components. The components have to be imported before they are rendered to string, and in order for them to use ``connect``, ``dispatch``, ``state``, or any other methods that you expose on the flux instances, they must reference an instance. In my app, as well as in your examples, this is done by importing a file that exports a flux instance, so it can be cached and used throughout. This pattern works well on the client but (without ugly hackery, which I'll come back to) is incompatible with creating a new instance of flux in the route-handler.

Using a single instance of flux on the server has three main issues:
1. Setting the instance state
2. Propagating the state to the components
3. Resetting in advance of another request 

##### 1) Setting the instance state
In lieu of setting the flux state directly I decided to simply dispatch an action called 'INIT' with the initial state as its payload and have each store set their respective state accordingly. This works well, but is related to another issue as I'll point out later.

##### 2) Propagating the state to the components
When ``renderToString`` is called and the component instances are created, their initial state is set by calling the specifier on the cached lookup of ``flux.state``. 

```js 
// decorator for React class
let { hooks, state } = this;
return Component =>
	class extends Component {
		constructor(...args) {
			super(...args);
			// Initial state
			let lastState = specifier(state);
			this.state = { [identifier]: lastState };
```

As is, the initial state wont reflect the changes made from the prior dispatch. My quick fix was to force the property lookup.

```js 
// decorator for React class
let { hooks, state } = this; // cache the property lookup
let self = this; // force the property lookup
  return Component =>
    class extends Component {
      constructor(...args) {
        super(...args);
        // Initial state

        // doesn't reflect changes to flux.state
        let lastState = specifier(state);

        // reflects changes to flux.state
        let initialState = specifier(self.state);

        // works
        this.state = { [identifier]: initialState };
```

##### 3) Resetting in advance of another request 
A [known issue](https://github.com/facebook/react/issues/3620) with React is that calling ``setState`` on the server causes an error. The first render avoids calling setState because the component is being initialized, and the setState hook hasn't been registered yet. The next request, however, triggers setState when it dispatches (INIT) the new state. My quick and dirty fix for this was to create an ``unhookAll`` method in flux.js which removes all hooks. A better solution would be to separately store setState registrations and user-defined hooks, and to expose a method to remove all the setStates. This method could also reset flux.state, like ``alt.flush`` does.

##### Ugly hackery

I thought of two ways of making server-side rendering while creating a new flux instance for each request. 

1) Wrap flux/index.js in a getter/setter.
Instead of returning a flux instance, the module would return getter/setter methods which would have to be used to access the instance. In order to ensure that you are always using the correct instance of flux, every single method call has to take the form ``flux.get()[method](...)``, which sucks. 

I use this pattern for maintaining a shared reference to an instance of react-router on client/server and it works well, partly because it is only ever referred to in a handful of places.

```js
var _router = null;
exports.set = router => _router = router;
exports.get = () => _router;
```

2) Pass flux instance to components
Instead of directly importing the flux instance into their modules, I could wrap every single component in a function that accepts a flux instance. This adds a great deal of complexity and also breaks the pattern of exporting a class per module.

**Comment by ameensol** [comment]
tl;dr ``fluxette`` is like 1.7 methods away from having awesome server-side rendering out of the box

**Comment by edge** [comment]
I'll give this a good read through. This looks great.

**Comment by edge** [comment]
I've been worming about isomorphic for a while, and my initial solution was dependency injection. That way, you can still import Stores and etc. as modules, but they would be functions, and you would get your own specific instance of flux. I haven't tested it yet, but it should work, in theory. The problem with that is, you get a bunch of overhead/boilerplate in wrapping your modules with functions.

The other solution would be to do it like redux and keep the state outside of the Stores completely, thus giving you *true* stateless Stores. But that changes the idea of a self-managed (i.e. quasi-stateless) Store function, which I really like.

As you can see, "make isomorphic easier" is on the todo list :)

**Comment by edge** [comment]
But my problem with doing it the redux way is that it adds these intrusive `<Provider />` components and stuff. I would be willing to consider that if it made everything else much easier.

**Comment by edge** [comment]
>A better solution would be to separately store setState registrations and user-defined hooks, and to expose a method to remove all the setStates. This method could also reset flux.state, like alt.flush does.

I think this might also pave the way for a further reduction in API surface area; managing hooks, middleware, state, etc. could be the responsibility of plugins. All `fluxette` then has to do is to make sure the actions are dispatched and the state is updated, calling any plugins that have hooked into the process along the way.

**Comment by ameensol** [comment]
Out of curiosity, why not recycle the flux instance? As a user it seems easier for me to call a method to reset state than it is to wrap all my components.

**Comment by edge** [comment]
Yeah, recycling would be free if we made Stores completely stateless and exported an initial state function or something.

**Comment by ameensol** [comment]
You might be able to keep Stores the way you have them now so long as the initial state function could force them to reset, right?

**Comment by edge** [comment]
I think, yeah, we could use your idea of an INIT action to set the state. Then recycling is free, *and* we don't have to modify the Store implementation.

**Comment by ameensol** [comment]
Using an INIT action does mean that each store has to implement a listener which copies its specific data from some initial state object, but I don't really see a way around that.

```js
  [INIT]: (user, action) => {
    return action.state.user
  },
```

**Comment by edge** [comment]
We could export an init constant that each store would automatically listen to and call the first argument (function) to get the state.

```js
export const init = '@@fluxetteInit';
```

```js
let store = Store(initAction => state, reducers);

dispatch({ type: init, data });
```

**Comment by edge** [comment]
This also gives us declarative rehydration, although we could probably find a way to make that easier. Maybe add a cursor to the dispatcher to automatically localize the data per store?

**Comment by edge** [comment]
After thinking about it, we could just skip all that with true stateless stores. It ultimately makes everything more stable. Although that would kill the idea of self-managed stores (good for non-React applications), I think it's a tradeoff that we could consider.

**Comment by ameensol** [comment]
So let me get this straight -- right now, each store manages its own state, which they update anytime they receive an action that matches one of their reducers, and once all stores have processed the action, flux.state is derived from the updated states of each of the stores.

This makes it so you can't just "reset" flux.state, because the stores are still holding on to their respective states via closure.

Making stores stateless would mean that anytime a store processes an action, it updates its branch of the global state tree, right? I did a little light reading around this and it seems like the two ways people optimize this are via cursors for synchronizing state tree branches with the global tree and using an immutable global state tree for fast reference equality checks in shouldComponentUpdate. It seems like both of these are worth considering, though I'm not sure what it would take to implement them.

As you mention a few comments up, using a default initAction reducer to process an init action would reduce the boilerplate involved in setting the initial state of the stores. It also follows the same pattern as setting component initial state in ``@connect``, and seems the easiest to implement. You might be able to make this easier via cursors, but you keep API surface area smaller by just using dispatch.

What are the other advantages to true stateless stores?

**Comment by ameensol** [comment]
Also, I'm pretty close to having filter.news ready to move to fluxette in production. Just need to force the flux.state property lookup in the component decorator and figure out a better way to get rid of all the setState hooks before each request. Resetting state by dispatching an action is fine, but nuking all the hooks by exposing this method in ``fluxette/lib/index.js`` isn't very pretty and only works because I'm not as of yet using any other hooks:
```js
let flux = new Fluxette(stores);
let api = {
  //...
  unhookAll: function () {
    while(flux.hooks.length) {
      api.unhook(flux.hooks[0])
    }
  }
}
return api         
```
I'm really impressed that you managed to get this whole thing working in 200 lines of code. Alt.js is near 2500 loc. You can fit your whole library into their index.js file.

**Comment by edge** [comment]
Haha, thanks.

You're right, making stores stateless means that the global state is passed as an argument to the Store, along with the action. That kind of reduces the loose-coupled-ness of stores, although that can be avoided by doing something like redux's composeStores or iterating through stores and the state in parallel, passing the corresponding branch to each store. I like both ideas, although we probably don't have to do either.

I think we'll stick with our original stores for now and use the initAction idea. One big problem in fluxette is that listening to changes above Stores will always result in a setState, because we create a new state object per dispatch. I talked to @gaearon about false positives in setStates, and I think using the techniques found in https://github.com/faassen/reselect would be a good idea. It was the better of three ideas that I thought might work, the other two being Store listeners and tree dependency management (we'll call it TDM). I don't want to do Store listeners because you can't listen to anything above a Store, and it would also change the way our stores work. TDM on the other hand, could be interesting.

TDM would work like this:
- create a snapshot of the current state
- iterate through each Store with the actions
- compare Store's state with previous state
- if the Store's state has changed, invalidate all the way up the tree

Invalidation would basically involve marking each the state as changed, and doing a
```js
stateBranch = { ...stateBranch };
```
Until we get to the root.

This would ensure that there aren't any false-negatives for strict equality checks, solving caching problems.

I still prefer the reselect method because it's efficient, but it has the most code smells. We'll work it out eventually.

In https://github.com/reactjs/react-rails/issues/146, @ericclemmons mentions that you can fix your server problem by moving the setState to `componentDidMount`. Apparently that got fixed in react 0.14 beta, but for now I can move it out from the constructor.

**Comment by edge** [comment]
One last thing I'm truly concerned about is the @connect decorator. We'll have to make it anonymous before we can get rid of any trace of singletons and be truly isomorphic. My problem with that is, how are we going access the instance of fluxette itself? Use context? props? I'm aware of the performance issues of context, and passing it down to each component as a prop sounds like it will have performance issues of its own, as well as the fact that we're probably going to need to intrude with some ugly FluxComponents to make that happen.

**Comment by ameensol** [comment]
I'm going to grok all this in a bit, but I wanted to make a quick point first. I read the react-rails issue you linked to, but the setState call isn't coming from a ``componentWillMount``. The setState is being called by the listener set in the ``@connect`` decorator when I first instantiate my components, which happens the first time I call ``renderToString`` on the server. So the very first time I make a request to the server, the component's setState is pushed to flux.hooks, but not called, as the state is set via ``this.state = ...`` in the constructor. The reason I get an error is that on the *next* request, when I dispatch my init action to set the flux state before calling ``renderToString``, the hook containing the component's setState is triggered. This is why I'm nuking all the hooks between requests. Again, this is only an issue because I'm trying to use a single instance of fluxette on the server, for reasons I stated in the original post.

**Comment by ameensol** [comment]
You're right that the **@connect** decorator is a weak point. It's really the only thing preventing me from creating a new instance of fluxette for every request.

**Comment by ameensol** [comment]
I read through reselect and I'm a fan. I like the pattern of listening on changes to the global tree, and reselect would make sure that a component is only re-rendered when the branch it cares about changes. 

I also think TDM would be easy to implement with your setup, especially because you recursively derive flux.state from your stores after every dispatch. 

Despite this, I agree that reselect would be more efficient than TDM. For a given higher-order component using reselect, since reselect would do deep equality checks on only the branches you care about, you would avoid doing those same calculations under TDM for all your stores. 

I'm curious -- what did you find smelly about reselect?

**Comment by edge** [comment]
The fact that we have to explicitly define dependencies is a bit smelly, since that should be an implementation detail. But I like how it allows you to compose selectors, which I think is a more than fair tradeoff.

We'll probably do it that way.

**Comment by edge** [comment]
I've been thinking a lot more about composeStores. Since fluxette is *really* unopinionated, any redux store could be plugged in if we just passed the state, so any redux-like feature would be a cakewalk to implement.

It's possible that we can have stores worry about their own listeners, calling them only if an action has been processed. That way it's more efficient, and we don't have to worry about selectors at all. It's a lot closer to the original concept of flux, in that stores *do* care about listeners, but, you know, the best of both worlds.

**Comment by edge** [comment]
I think I've settled on `context` for passing around the flux object. Since that's kind of what it's for.

**Comment by edge** [comment]
Turns out that a twelve hour flight without wifi does wonders to your productivity.

I've basically solved all the problems on the `context-ssr` and `store-composition` branches.

**Comment by edge** [comment]
I'm pretty comfortable with sticking to all the changes in `context-ssr`, but I think we'd have to drop the "minimalist" if we used `store-composition`. It doesn't really let you compose stores yet, but you can declaratively rehydrate your state now. It uses cursors and stuff. The initAction thing works more or less perfectly.

Again, I think we could work actual store composition and the "minimalist" title back in if we used stateless stores.

**Comment by ameensol** [comment]
Great work! 

Passing the flux instance through context seems to allow for an easier isomorphic setup. Since connect and link are both anonymous, we don't need the flux instance when importing the components. I was worried that we might need to lazy-load the components, but this setup conveniently avoids that too.

Something that doesn't feel quite right in the ``store-composition`` branch is how the rehydration works. 

```js
let { type } = action;
  if (type === initType) {
  // Rehydrate
  state = action.state !== undefined
    ? action.state : statefn();
}
```

In my usage, since each store is managing its own state, I pass a reducer to each store that listens for the initAction and copies the relevant properties from the global state into the store's state. For example, in the user store I posted before, I copy only the user property:

```js
[INIT]: (user, action) => {
  return action.state.user
}
```
As is, it seems like your rehydration code will only allow stores to copy the global state. Perhaps we could change the rehydration code to:
```js
let { type } = action;
  if (type === initType) {
  // Rehydrate
  state = action.state !== undefined
    ? statefn(action.state) : statefn();
}
```
So ``storefn`` would optionally accept the global state and have the following signature:
1. When called **without** state, return the initial state for the store (e.g. {})
2. When called **with** state, return the branch of the global state that we want to set to the state of the store (e.g. state.user)

There actually might be a better way of handling this internally since I'm pretty sure the structure of the global state tree is analogous to the stores object:
```js
stores: { user: @Store, topics: @Store }
state: { user: { /*data*/ }, topics: { /*data*/} }

**Comment by ameensol** [comment]
On a second pass, it does seem like the complication that this introduces could be resolved using stateless stores...

**Comment by edge** [comment]
Could you work out a prototype? I'm working on stateless stores right now, and I really want it to make it the smoothest possible experience. Doing it the redux way basically requires adding like two lines, but if there's a better way, I'm all for it.

**Comment by edge** [comment]
I'm also in a place where the internet is really spotty, so I won't be able to get on much.

**Comment by edge** [comment]
Ended up with this *beautiful* one line dispatcher.

```js
flux.state = actions.reduce(flux.store, flux.state);
```

**Comment by edge** [comment]
This is becoming the most beautiful thing ever.

**Comment by ameensol** [comment]
Pretty indeed.

And sure, I'll give stateless stores some thought. 

Also, what is the state initialized to if ``auto`` is false?

**Comment by ameensol** [comment]
So I thought about it today but I didn't think of anything simpler than how you have it -- treat flux.store as a top-level function that takes the current state and an action, and returns an updated state.

flux.store in this case is really a container for "stores" which are themselves just really containers for "reducers", which are the functions that return updated state for each specific action.

But since each store can be invoked with the global state, the logical separation of stores is up to the user. Complex operations that involve data from multiple stores are trivial to implement. 

This is great.

**Comment by edge** [comment]
Yep, that's what I was onto last night (day).

I'll push changes once tests pass.

**Comment by edge** [comment]
Now there are two reducer factories -- `Store` and `Reducer`. Consider our state a tree. `Store`s are the branches. They don't do reduction on their own, but manage `Reducer`s. These are the leaves of the tree, and basically what we used to call "Stores". It makes everything a lot easier.

**Comment by ameensol** [comment]
Cool, I'm excited to see what you've got in the Interface. 

Not sure if you want to keep these in userland or include them for convenience, but I wrote some helper functions to auto-generate types and action creators from an array of my API client method names. 

**types.js**
```js
import { methods } from '../utils/api';

// returns object w/ REQ/FAIL/DONE types
var createType = (method) => ({
  REQUEST: 'API_'+method.toUpperCase()+'_REQUEST',
  FAIL: 'API_'+method.toUpperCase()+'_FAIL',
  DONE: 'API_'+method.toUpperCase()+'_DONE'
})

// returns an object of action types, one for each action method
var createTypes = (methods) => {
  var API =  methods.reduce((types, method) => {
    types[method.toUpperCase()] = createType(method)
    return types
  }, {})
  API.QUERY = 'API_QUERY'
  return API
}

export default {
  INIT: 'INIT',
  API: createTypes(methods)
}
```

**api.js**
```js
import { Mapware } from 'fluxette'
import { API } from '../types';
import { dispatch, proxy } from '..';
import api from '../../utils/api';
import { methods } from '../../utils/api';

var createCreator = (method, API) => ({
  request: (data) => ({ type: API[method.toUpperCase()].REQUEST, data }),
  fail: (data) => ({ type: API[method.toUpperCase()].FAIL, data }),
  done: (data) => ({ type: API[method.toUpperCase()].DONE, data })
})

// returns an object of action creators, one for action
// API is an object containing the type definitions for each method
var createCreators = (methods, API) => {
  return methods.reduce((creators, method) => {
    creators[method] = createCreator(method, API)
    return creators
  }, {})
}

var creators = createCreators(methods, API)

creators.query = (method, data) => ({ type: API.QUERY, query: { method, data } })

const ware = Mapware({
    [API.QUERY]: action => {
      let { method, data } = action.query
      // 1 launch async request
      api[method](data, (err, res, body) => {
        if (err) {
          dispatch(creators[method].fail({ ...data, err }));
        } else {
          dispatch(creators[method].done({ ...data, body }));
        }
      })
      // 2 return modified action -- use creator
      return creators[method].request(data)
    }
})

proxy(ware)

export default creators
```
I took you advice and am handling async data-fetching as middleware, but for additional simplicity I'm using the ``API.QUERY`` type as a generic proxy for all my async dispatches. 
```js
dispatch(actions.api.query('getArticles', { id: this.props.topic.id }))
```
I like this because it also allows me to directly call my api client methods and dispatch my actions should I need to, but it also allows me to fire and forget the vast majority of the time.

**Comment by ameensol** [comment]
This optimization coupled with more concise store code and I actually removed ~500 lines by switching over.

**Comment by edge** [comment]
Ah. I ended up deciding to remove middleware completely, allowing you to instead to override `dispatch` on `Interface`. I did so because there ended up being a lot of ways to implement middleware (per action? per dispatch? mutations vs. side effects? etc.)

But I'll take a look at what you've got there.

**interface.js**
```js
import { normalize, deleteFrom } from './util';

export default class {
	constructor(flux) {
		this.instance = flux;
	}
	dispatch(...actions) {
		let { instance } = this;
		instance.process(normalize(actions));
	}
	process(...args) {
		this.instance.process(...args);
	}
	init(state) {
		let { instance } = this;
		instance.history = [];
		instance.state = state !== undefined ? state : instance.store();
	}
	state() {
		return this.instance.state;
	}
	history() {
		return this.instance.history;
	}
	hook(...args) {
		this.instance.hooks.push(...args);
	}
	unhook(fn) {
		deleteFrom(this.instance.hooks, fn);
	}
}
```

**Comment by ameensol** [comment]
Looks clean. I like it. I'll resentfully re-write the middleware once you've pushed.

**Comment by edge** [comment]
Sorry. Yeah, it looks like 0.13.x (1.0.0-rc?) will have huge breaking changes.

**Comment by ameensol** [comment]
No worries, was only joking. Glad I got to help with the debugging :)

**Comment by edge** [comment]
Pushing soon. The docs are going to need a total rewrite, haha.

**Comment by edge** [comment]
@ameensol Just merged. Check out the tests while I write some docs.

**Comment by edge** [comment]
Closing, because 0.14.x basically solves all the problems here.
